### DISCUSSION TITLE
Request for Comments: Display of Workflow Test Results

### DISCUSSION THREAD

**Discussion by hexylena** [galaxyproject/training-material#4151] [Ideas]
The GTN is finally implementing automated workflow testing for the workflows across the GTN. We have started the implementation, but we have some questions for you, **the community**! Namely, how should we present these results to you?

**Q1**: What information is useful?

- we'd have possibly 0-3 results from the 'big three' servers.
- it'll include whether the tests passed or failed (could be displayed binary :heavy_check_mark: :x:  or as % or something else)
- We will link to this "test history" that will be made public
- we also have more information like:
    - what were the input files (location on zenodo, filetype)
    - for each step in the workflow
        - Step Name
        - how much time it took to run (individual steps + time between / end-to-end time)
        - memory/cpu/etc metrics.
        - command line, stdout/stderr (often empty)
        - job parameters
        - tool id

**Q2**: Where would you like to see it?

- We can display it on the individual workflow (as a tiny lil button)
- or a separate page for all workflows for a tutorial
- or a page per workflow!
- Or on a page summarising all workflows for a topic, or the whole GTN!
- (or all of the above!)

---

We would appreciate you making **individual reply posts** below, something like:

> As a **[learner/tutorial author/teacher/server admin/etc]**, I would find it useful to see ... and ...

So we can group use cases :) Please feel free to also add your response to an existing reply, if you feel like it fits in the same category as theirs.

**Comment by nomadscientist** [comment]
Oh there's a way to have multiple workflows in the tutorial? Thank god, I had concerns.
I think it should be a button you click by the workflow on the tutorial that takes you to a second page with ALL the workflow runs, but then a scrolling tab list (in case there are a lot) at the left of that second page.

As a learner and trainer, I would find it useful to see the timestamps (for how long a tool took) by the tools in the tutorials themselves, to indicate how long I should expect it to take.

**Reply by hexylena** [reply]
Would you be interested in historic run data then? in addition to just a "latest run"?

**Reply by hexylena** [reply]
and yes, e.g. [16s short has a lot of workflows](https://training.galaxyproject.org/training-material/topics/metagenomics/tutorials/mothur-miseq-sop-short/workflows/), hence the display/location of such data being an important question.

**Reply by nomadscientist** [reply]
Yes historic data would be ace!

**Comment by mira-miracoli** [comment]
As a server admin, I would find it useful to see everything listed in Q1 if possible, priority would be link to test history, metrics and file I think.
Display: one list with :heavy_check_mark: / :x:  and a page with all information linked there.

Thank you for doing this!

**Reply by hexylena** [reply]
this is very helpful, thank you :yellow_heart:

**Comment by yvanlebras** [comment]
As a tutorial author/teacher, it seems to me of interest to have all points listed in Q1, with a particular emphasis for 1/ time used by each tool execution and 2/ access to an existing history created by the workflow (at least to have possibility to reuse existing good results if there is issues regarding tool execution or tool schedulong at the moment of a trinaing).

Maybe display it on the tuto page on a workflow oriented      part with maybe a possibility to scroll down elements can be top!

AMAZING work !!!! Thank you !!!!

**Reply by hexylena** [reply]
Thanks @yvanlebras great to know! I suspected the tool run times would be useful, thank you for confirming it and providing some suggestions :)

**Comment by ilveroluca** [comment]
As a tutorial author and [LifeMonitor](https://www.lifemonitor.eu/) developer (with @simleo and @kikkomep), I would like to see these test outcomes on a [LifeMonitor dashboard](https://www.lifemonitor.eu/lm_dashboard).  One of the LifeMonitor service’s goals is to collect in a single place test outcomes from multiple workflows of interest, potentially running on multiple infrastructures (we call them “test instances"). Once a workflow test is registered, interested people can subscribe to see it on their own dashboards and/or receive notifications about its status.  Test results can also be retrieved via REST API for easy integration. There’s also a GitHub app that can help keep LifeMonitor (and WorkflowHub) entries up-to-date.

This is already in place for the [IWC workflows](https://github.com/galaxyproject/iwc) (we can thank @simleo for that).  Naturally, if there is interest in registering these workflows in LifeMonitor, we’d be happy to discuss the details to come up with a strategy.

**Reply by hexylena** [reply]
Hey @ilveroluca that's great to hear about. I hadn't had time to fully go through the lifemonitor tutorial but I clearly need to now! 

We've been looking for a way to keep the GTN workflows updated on Workflow Hub for some time now (#3049) but hadn't found anything acceptable, since all of the solutions required us to manually create all 200+ entries in WFH.

We won't be able to change the repository layout significantly so, looking at the documentation, we'd need to write some automation around generating the WTROCs ourselves, and then sending them to LifeMonitor? Is that correct? 

Additionally how we'll launch tests is maybe a bit complicated since we're trying to do some special things like, launching the workflows in one github action, and collecting the results in a separate one, as we don't want to burn github action minutes while some of these workflows take multiple hours to run against EU/Org.

And that's our third oddity, we're testing workflows against multiple backend servers. I'm not sure if that sort of workflow is supported? 

We should definitely discuss the best solution here since in general I like the suggestion of also pushing the results to LM, though we'll probably continue to develop some form of internal dashboard given different audience needs

**Reply by ilveroluca** [reply]
Hi @hexylena!

We reasoned a bit about the information you provided about the GTN setup.  I think we can more or less split the discussion over two main points:  test monitoring and automated creation/maintenance of workflow entries on WorkflowHub and LifeMonitor.

## First point: test monitoring

From how we interpreted your comments, it seems that the GTN workflow testing setup is not implemented yet, and it we hypothesize that the plan is to have an action that launches multiple tests on one or more (maybe public) Galaxy services and have a separate *single* action to wait and collect all test results. This last point doesn't work with LifeMonitor's model (but please correct us if we didn't understand your plan).

LM doesn't know about Planemo or any other test engines.  What it currently does is look at the success or failure of a CI workflow. It thus follows that there must be one action per workflow. Maybe you can find a way to have your single waiting action dispatch other workflow-specific actions to collect and propagate test results (e.g., GH Actions allows one workflow to call other workflows, but [the limitations](https://docs.github.com/en/actions/using-workflows/reusing-workflows#limitations) seem to get in the way of using this approach for 200+ workflows).  On the other hand, monitoring multiple testing services for the same workflow is not a problem: LifeMonitor supports the concept of having multiple test instances for the same workflow. These test instances can be on different servers and they can even be different type of CI services (e.g., the same tests running on both Jenkins and GH Actions).


## Second point: automated creation/maintenance of workflow entries on WorkflowHub and LifeMonitor

This functionality is supported through the [LifeMonitor GitHub app](https://github.com/apps/lifemonitor). It is indeed a more complex feature to support in the GTN case, since at the moment LM makes the assumption of one workflow per repository (and RO-Crates in a way also make this assumption, as the metadata file is supposed to be at the root of the RO-Crate archive and WROC represents a single workflow).  To enable automatic registration we have to be able to automatically create a WTROC for the workflow. For this purpose, we have developed the [repo2crate](https://github.com/crs4/repo2rocrate/) package, which LifeMonitor uses.  Currently, LM executes repo2crate *at the base of the repository* and uses the resulting WTRO-Crate to register or update the workflow's entries on LifeMonitor and/or WorkflowHub.  Note that you may also find repo2crate and [ro-crate-py](https://github.com/ResearchObject/ro-crate-py) useful for any RO-Crate-related custom automation.

We reasoned that LifeMonitor should in theory be able to support a multi-workflow repository such as GTN's by associating a "prefix" path to the workflow to select a subdirectory of the repository to be considered as its "root path".  This would require some development work on our part. Also, the end user would have to identify each workflow and the associated prefix path *and the associated test instance* through the [lifemonitor.yml](https://www.lifemonitor.eu/lm_wft_best_practices_github_app#configuration-file) configuration file (through the creation could be partly automated).  At this stage we're not sure when such a feature could be integrated, though interest from the GTN would be good motivation.

A final thing I should mention is that both LifeMonitor and WorkflowHub provide REST APIs ([LifeMonitor API explorer](https://api.lifemonitor.eu/static/specs/apidocs.html), [Workflowhub API](https://workflowhub.eu/api), [WorkflowHub RO-Crate submission API](https://about.workflowhub.eu/developer/ro-crate-api/)).  You may also consider targeting these directly in your automation workflow.

**Comment by nomadscientist** [comment]
I'd like a summary not just by tutorial but by theme (i.e. single cell) so communities know who to chase / check in on
